pinot-common/src/main/java/org/apache/pinot/common/utils/request/RequestUtils.java

```diff
@@ -107,7 +107,11 @@ public static Expression getLiteralExpression(SqlLiteral node) {
     Expression expression = new Expression(ExpressionType.LITERAL);
     Literal literal = new Literal();
     if (node instanceof SqlNumericLiteral) {
-      if (((SqlNumericLiteral) node).isInteger()) {
+      // Mitigate calcite NPE bug, we need to check if SqlNumericLiteral.getScale() is null before calling
+      // SqlNumericLiteral.isInteger(). TODO: Undo this fix once a Calcite release that contains CALCITE-4199 is
+      // available and Pinot has been upgraded to use such a release.
+      SqlNumericLiteral sqlNumericLiteral = (SqlNumericLiteral) node;
+      if (sqlNumericLiteral.getScale() != null && sqlNumericLiteral.isInteger()) {
```

Comment:
Will this condition ever evaluate to true assuming scale will be non null for decimal values and isInteger will be false ?

Reply:
May be it will evaluate to true for values like `<something.0>`

Wondering why we never hit this earlier. scale will always be null for a big decimal value that has nothing to the right of decimal point. So `isInteger()` should have always thrown NPE in the calcite version we are using

So if the SqlNumericLiteral node has following values

123456.0 -> non null scale -> scale value  0 -> casted to integer
123456789 -> null scale value -> should have thrown NPE
12345 -> null scale value -> should have thrown NPE
1234567.787878  non null scale -> scale value  non 0 -> casted to double in the else branch

So not sure why 2nd and 3rd case didn't throw NPE all this while

Reply:
Apparently, this happens when Calcite Parser thinks a literal is:
```
    <APPROX_NUMERIC_LITERAL> {
        return SqlLiteral.createApproxNumeric(token.image, getPos());
    }
```
where an APPROX_NUMERIC_LITERAL is any numeric value with an exponent:
```
< APPROX_NUMERIC_LITERAL:
    (<UNSIGNED_INTEGER_LITERAL> | <DECIMAL_NUMERIC_LITERAL>) <EXPONENT> >
```

From what I can tell, `SqlLiteral.java` class in Calcite needs to be fixed as well (but the bug won't appear in their master branch since they have already added a null check). I will ping their mailing list and see if this is something that should be fixed.

```
 CURRENT:
  public static SqlNumericLiteral createApproxNumeric(String s, SqlParserPos pos) {
    BigDecimal value = SqlParserUtil.parseDecimal(s);
    return new SqlNumericLiteral(value, (Integer)null, (Integer)null, false, pos);
  }
  
 FIXED:
  public static SqlNumericLiteral createApproxNumeric(String s, SqlParserPos pos) {
    BigDecimal value = SqlParserUtil.parseDecimal(s);
    return new SqlNumericLiteral(value, value.precision(), value.scale(), false, pos);
  }

```
But it will probably take a while for these fixes to show up in their release, so we should still mitigate the issue in Pinot.